integrations/tensorflow/iree_tf_compiler/TF/SavedModelToIreeABI.cpp

```diff
@@ -0,0 +1,577 @@
+  SymbolTable::setSymbolVisibility(wrapperFunc,
+                                   SymbolTable::Visibility::Public);
+  Block *entryBlock = wrapperFunc.addEntryBlock();
+  builder.setInsertionPointToEnd(entryBlock);
+
+  // Flatten the arguments.
+  // For each argument of the wrapper function, associate with a
+  // StructureLevel and recursively emit dereferencing ops until reaching a
+  // leaf.
+  SmallVector<Value> callArgs;
+  callArgs.resize(internalFunc.getNumArguments());
+  for (auto it : llvm::enumerate(argsRoot.children)) {
+    BlockArgument wrapperArgValue = entryBlock->getArgument(it.index());
+    it.value().emitDereferenceArgs(loc, builder, wrapperArgValue, callArgs);
+    refArgs.push_back(it.value().createReflectionType());
+  }
+  assert(llvm::all_of(callArgs, [](Value v) { return v != nullptr; }) &&
+         "not all call arguments mapped");
+
+  // Emit the call to the internal func.
+  ResultRange internalResults =
+      builder
+          .create<CallOp>(loc, internalFunc.getType().getResults(),
+                          internalFunc.getName(), callArgs)
+          .getResults();
+
+  // And then unflatten the results for return from the wrapper.
+  SmallVector<Value> wrapperReturns;
+  wrapperReturns.resize(wrapperResultTypes.size());
+  if (resultsRoot.type == LevelType::None) {
+    assert(wrapperReturns.empty() && "mismatched none return");
+  } else if (isMultiResult) {
+    // Multiple return.
+    assert(resultsRoot.children.size() == wrapperReturns.size() &&
+           "mismatched multiple result arity");
+    for (int i = 0, e = resultsRoot.children.size(); i < e; ++i) {
+      wrapperReturns[i] = resultsRoot.children[i].emitCreateReturns(
+          loc, builder, internalResults);
+      refReturns.push_back(resultsRoot.children[i].createReflectionType());
+    }
+  } else {
+    // Single return.
+    assert(wrapperReturns.size() == 1 &&
+           "mismatched return arity for unary func");
+    wrapperReturns[0] =
+        resultsRoot.emitCreateReturns(loc, builder, internalResults);
+    refReturns.push_back(resultsRoot.createReflectionType());
+  }
+
+  assert(llvm::all_of(wrapperReturns, [](Value v) { return v != nullptr; }) &&
+         "not all call returns mapped");
+  builder.create<ReturnOp>(loc, wrapperReturns);
+
+  // Add reflection attribute.
+  std::string refStr;
+  {
+    json::Object refDict;
+    refDict["a"] = json::Value(std::move(refArgs));
+    refDict["r"] = json::Value(std::move(refReturns));
+    json::Value refDictValue(std::move(refDict));
+    llvm::raw_string_ostream refOut(refStr);
+    refOut << refDictValue;
+    refOut.flush();
+  }
+
+  SmallVector<NamedAttribute> funcReflectAttrs(
+      {builder.getNamedAttr("abi", builder.getStringAttr("ireevm")),
+       builder.getNamedAttr("abiv", builder.getI32IntegerAttr(1)),
+       builder.getNamedAttr("d", builder.getStringAttr(refStr))});
+  wrapperFunc->setAttr("iree.reflection",
```

Comment:
It'd be nice here to do this as a dialect attribute instead - `#iree.abi<...>` or whatnot - I found it tricky to work with `iree.reflection` as it's kind of a grab-bag for everything and the merging/combining code for SIP stuff deeper down in the system would always trip me up. We could use an AttributeInterface as a trait marker for allowing attributes to be serialized to the VM. Filed #5543 to track.